execgen: add generic inliner

jordanlewis opened the pull request:
This commit adds a new directive to execgen:

// execgen:inline

This directive causes a function that's annotated with it to be inlined
into all of its callers via AST transformation. This kind of inlining is
smarter than before - it permits arguments with names different from the
names of the function's formal parameters, for example.

This commit also changes the functions in distinct_tmpl to use this
directive instead of the manual templating method.

I think this is the only easy function to change. The rest have, at the
simple level of difficulty, static template-time parameters, and at a
harder level of difficulty, type parameters. Improving these cases
holistically should come next.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49728)
<!-- Reviewable:end -->

jordanlewis committed:
colexec: don't randomize batch size in benches

This was making it hard to run the same benchmarks in a consistent way.

Release note: None

jordanlewis committed:
execgen: add generic inliner

This commit adds a new directive to execgen:

// execgen:inline

This directive causes a function that's annotated with it to be inlined
into all of its callers via AST transformation. This kind of inlining is
smarter than before - it permits arguments with names different from the
names of the function's formal parameters, for example.

This commit also changes the functions in distinct_tmpl to use this
directive instead of the manual templating method.

I think this is the only easy function to change. The rest have, at the
simple level of difficulty, static template-time parameters, and at a
harder level of difficulty, type parameters. Improving these cases
holistically should come next.

Release note: None

rohany commented on pkg/sql/colexec/execgen/cmd/execgen/inline.go:
```diff
@@ -0,0 +1,252 @@
+		case *dst.FuncDecl:
+			var mustInline bool
+			for _, dec := range n.Decorations().Start.All() {
+				if dec == "// execgen:inline" {
```
huh, i thought it would have a smarter way of parsing out these annotations

rohany commented on pkg/sql/colexec/execgen/cmd/execgen/inline.go:
```diff
@@ -0,0 +1,252 @@
+				// Nothing to do, but recurse further.
+				return true
+			}
+			for _, p := range n.Type.Params.List {
```
What is this checking?

rohany commented on pkg/sql/colexec/execgen/cmd/execgen/inline.go:
```diff
@@ -0,0 +1,252 @@
+				}
+			}
+			templateFuncMap[n.Name.Name] = n
+			cursor.Replace(&dst.GenDecl{
```
yeah this needs some comments, I can't really tell what it's doing

rohany commented on pkg/sql/colexec/execgen/cmd/execgen/inline.go:
```diff
@@ -0,0 +1,252 @@
+			// convert the result into the assignment variable.
+
+			// Generate return value variables.
+			results := decl.Type.Results.List
```
Hah, this is really a throwback to a compilers course in college

rohany commented on pkg/sql/colexec/execgen/cmd/execgen/inline.go:
```diff
@@ -0,0 +1,252 @@
+				}
+				n := cursor.Node()
+				switch n := n.(type) {
+				case *dst.FuncLit:
```
Why is this not a `CallExpr` as well -- or is `FuncLit` something else?

rohany reviewed (commented):
Pretty slick -- do do you have a snippet of what the generated callsites look like now?

jordanlewis commented:
https://gist.github.com/jordanlewis/36b2cf775198886c6a65ac91de1911d0

jordanlewis commented:
Alright, redid a lot of this - added plentiful comments, moved files into `pkg/sql/colexec/execgen` instead of `cmd/execgen`, added unit tests for a couple of functions, and added a datadriven test that works on Go files with execgen annotations.

jordanlewis reviewed (commented):
test

rohany commented on pkg/sql/colexec/execgen/inline.go:
```diff
@@ -0,0 +1,363 @@
+	"go/parser"
+	"go/token"
+
+	"github.com/cockroachdb/cockroach/pkg/testutils/lint/passes/fmtsafe/testdata/src/github.com/cockroachdb/errors"
```
Is this really the package import -- this is like 150 characters long

jordanlewis replied:
lol, wtf. fixed.

rohany commented on pkg/sql/colexec/execgen/inline.go:
```diff
@@ -0,0 +1,363 @@
+		// and ExprStmt, which are simply:
+		// foo()
+		// AssignStmts need to do extra work for inlining, because we have to
+		// simulate producing return values
```
[nit]: missing period

rohany commented on pkg/sql/colexec/execgen/inline.go:
```diff
@@ -0,0 +1,363 @@
+			// function to be deleted, such as template comments like {{end}}. This
+			// is kind of a quirk of the way the comments are parsed, but nonetheless
+			// this is an easy fix so we'll leave it for now.
+			cursor.Replace(&dst.GenDecl{
```
Just to understand, this emits a "_ = `inlined_func`" where the function call was?

jordanlewis replied:
Yeah. Added a comment.

rohany commented on pkg/sql/colexec/execgen/inline.go:
```diff
@@ -0,0 +1,363 @@
+		}
+		n := cursor.Node()
+		switch t := n.(type) {
+		case *dst.FuncLit:
```
Can you add what go code matches a `FuncLit`?

jordanlewis replied:
Done, it's a function literal, like `x := func() bool {return true}`

rohany commented on pkg/sql/colexec/execgen/inline_test.go:
```diff
@@ -0,0 +1,102 @@
+		{
+			funcDecl: `func a(a int, b int) {}`,
+			callExpr: `a(x, y)`,
+			expectedStmts: `var (
```
What's the reason we do `var x <type> = y`, rather than `x := y`? We have a fresh scope right?

jordanlewis replied:
It's just for cleanliness, I dunno. I like the way a var () decl groups the related assignments together. But it's totally equivalent to n `x := y` statements.

rohany replied:
True -- no difference.

rohany commented on pkg/sql/colexec/execgen/testdata/inline:
```diff
@@ -0,0 +1,146 @@
+inline
```
cool test!

rohany reviewed (commented):
Thanks for all the comments, it helped alot! A few questions:
* This is a decent amount of code, and it helps us inline only this simple case as you mentioned in slack -- what is missing here, and how much code do you forsee needing to add for the more complicated cases?
* Will introduction of trying to template stuff affect this, or will it just be another pass (probably scheduled before this one)
* I'm worried about execgen becoming more of a "arcane" piece of code than it is now, with the introduction of the `dst` package -- it's like a mini DSL for go ast manipulation. Maybe that is an unfounded worry though.

jordanlewis commented:
Thanks for the review!

> This is a decent amount of code, and it helps us inline only this simple case as you mentioned in slack -- what is missing here, and how much code do you forsee needing to add for the more complicated cases?

Yeah, I'm definitely cognizant of how much effort this is. I guess I'm still kind of in an exploratory mode with this work. I'm not completely sure what the end goal is yet.

The next step I'd like to do is templating simple boolean arguments, which seems to be a common thing throughout the exec codebase. For example:

```
func a(i int) {
   if col.MaybeHasNulls() {
     x = stuff(i, true)
   } else {
     x = stuff(i, false)
   }
}

// execgen:inline
// execgen:template<hasNulls>
func stuff(runtimeArg int, hasNulls bool) int {
   if hasNulls {
      return 3
   } else {
      return 4
   }
}
```

Currently, this is done with a mix of text/template and regex which although probably less code, is pretty confusing, hacky, and hard to get started with.

Once that's done, the next step would be permitting "mid-stack inlining", so that you can pass template boolean arguments through more than one call site - this is also a common ask. This part seems annoying - you have to keep a graph of functions that are calling each other. It's definitely dawning on me that this is why you have IRs, but I don't think I want to get into messing with any non-AST IRs - if it comes to that I think this project will have jumped the shark.

Once that's done, the next step would be permitting non-boolean template arguments: specifically, `types.T` template arguments. If that part can be figured out, we'll be able to delete 90% of the custom, per-file template code.

The downside of the code that I'm writing is that inevitably it'll have some subtleties and restrictions - like I'm not planning to try to deal with having conditional expressions that mix template and runtime parameters, the engine will probably just ignore those - so that might be confusing too. I just think it'll be less confusing on the whole, and easier to get started with. 

> Will introduction of trying to template stuff affect this, or will it just be another pass (probably scheduled before this one)

It'll be a concurrent pass. As we're collecting the functions, we'll also note their template arguments down. As we're inlining the functions, we'll evaluate the template-time conditionals to only include the matched code in the inlined output.

> I'm worried about execgen becoming more of a "arcane" piece of code than it is now, with the introduction of the dst package -- it's like a mini DSL for go ast manipulation. Maybe that is an unfounded worry though.

It's definitely a founded worry. But try to take a look at some of the other code we already have. It's very, very arcane, and you have to be arcane in subtle different ways for every operator you write, right now. I'd much prefer to have a single arcane execgen than 100 arcane templates. Also, I hear your point about this becoming arcane, but hopefully the datadriven test styles I've added will make it pretty easy to test and understand.

jordanlewis commented:
This one guy has some prior art: https://github.com/cosmos72/gomacro/blob/master/doc/generics-c++.md

rohany commented:
>  It's definitely dawning on me that this is why you have IRs, but I don't think I want to get into messing with any non-AST IRs - if it comes to that I think this project will have jumped the shark.

Yeah, I'm unsure how far you can get without fancier IR's, but you can also just mock out some IR's with maps of AST nodes to semantic info.

> It'll be a concurrent pass. As we're collecting the functions, we'll also note their template arguments down. As we're inlining the functions, we'll evaluate the template-time conditionals to only include the matched code in the inlined output.

Are you sure? It feels to me like inlining and templates are different topics. The templating pass would create the templated versions of these functions, and mark them as inlineable, and then the inline pass goes and inlines them. Im imagining generating two `stuff` implementations (from above) with inline annotations, replacing each call with the correct one.

> Once that's done, the next step would be permitting "mid-stack inlining", so that you can pass template boolean arguments through more than one call site - this is also a common ask. This part seems annoying - you have to keep a graph of functions that are calling each other. 

I feel like this will be simpler if inlining and templating are considered separately, but I don't have concrete explanation why. 

> I'd much prefer to have a single arcane execgen than 100 arcane templates.

Heavily agree, I find the templates difficult to navigate.

yuzefovich reviewed (commented):
Thanks for pushing forward with this work!

How is the speed of generating code using this new approach in comparison to the old one?

> you have IRs

What is "IRs"?

> Reviewed 1 of 1 files at r1, 5 of 7 files at r2, 6 of 8 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-:-M99HMc684786CyBLYfY:b-ggdds0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

___
*[pkg/sql/colexec/main_test.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99HcQWCWAD0Aolcwmh:-M99HcQWCWAD0Aolcwmi:baji11d) ([raw file](https://github.com/cockroachdb/cockroach/blob/13845e418807a3a1afe290636937f5ebb3f1e81c/pkg/sql/colexec/main_test.go#L71)):*
> ```Go
> 
> 		flag.Parse()
> 		if f := flag.Lookup("test.bench"); f == nil || f.Value.String() == "" {
> ```

I was wondering about how one would do this at some point but ended up setting an environment variable to `false` on my machine. Turns out it was this simple, very nice change, thanks!
___
*[pkg/sql/colexec/execgen/inline.go, line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99LGZYEX4lYHgfvh8N:-M99LGZYEX4lYHgfvh8O:b4tfr6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline.go#L118)):*
> ```Go
> 				}
> 				// Replace the return with the new assignments.
> 				return &dst.BlockStmt{List: returnAssignmentSpecs}
> ```

I wonder why it is a block stmt, does it have to be so? Just curious. Or it doesn't matter because we're using only `body.List` below?
___
*[pkg/sql/colexec/execgen/inline.go, line 228 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99NFyK6nT7W9dzO7Yw:-M99NFyK6nT7W9dzO7Yx:b-g6yp38) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline.go#L228)):*
> ```Go
> 
> // extractReturnValues generates return value variables. It will produce one
> // statement per return return value of the input FuncDecl. For example, for
> ```

nit: `s/return return/return/`.
___
*[pkg/sql/colexec/execgen/inline.go, line 256 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99NgCf7_ebCCeeHtjG:-M99NgCf7_ebCCeeHtjH:b5l3lem) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline.go#L256)):*
> ```Go
> 		}
> 	}
> 	retVal := &dst.DeclStmt{
> ```

nit: `s/retVal/retValDeclStmt/`.
___
*[pkg/sql/colexec/execgen/inline.go, line 290 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99O2O-5IzpsIXvldsh:-M99O2O-5IzpsIXvldsi:boefpkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline.go#L290)):*
> ```Go
> 	reassignmentSpecs := make([]dst.Spec, 0, len(formalParams))
> 	for i, formalParam := range formalParams {
> 		if inputIdent, ok := callExpr.Args[i].(*dst.Ident); ok && inputIdent.Name == formalParam.Names[0].Name {
> ```

Just curious why we have `Names[0]`, can an AST node have multiple names somehow?
___
*[pkg/sql/colexec/execgen/inline_test.go, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99VjaK3q9POAXYAyqr:-M99VjaK3q9POAXYAyqs:bssmngs) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline_test.go#L72)):*
> ```Go
> 			decl: "func foo(a int) (int, string) {}",
> 			expectedRetDecls: `var (
> 		__retval_0 int
> ```

super nit: I'd indent these `expectedRetDecls` a little bit to the right, if that's ok.
___
*[pkg/sql/colexec/execgen/cmd/execgen/main.go, line 134 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99WS2048j-Tgl6-Ln5:-M99WS2048j-Tgl6-Ln6:b95dkek) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/cmd/execgen/main.go#L134)):*
> ```Go
> 			return err
> 		}
> ```

nit: seems like unnecessary newline.
___
*[pkg/sql/colexec/execgen/testdata/inline, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M96KdvITnxKnqOEXInJ-r3-1:-M99Wr1X8ynTt0WEKFjp:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/303646c451114b54361b8e07e43d9971e8055794/pkg/sql/colexec/execgen/testdata/inline#L1)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

cool test!
</blockquote></details>

+1
___
*[pkg/sql/colexec/execgen/testdata/inline, line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99WcFtA8L67XNSfwhB:-M99WcFtA8L67XNSfwhC:b-da4wuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/testdata/inline#L21)):*
> ```
> 
> func a() {
> 	{
> ```

nit: the indentation here seems to be done with tabs, can we adjust that? I don't have a personal opinion on tabs vs spaces, but I think in our codebase we use spaces.


<!-- Sent from Reviewable.io -->

jordanlewis committed:
colexec: remove unnecessary execgen imports

The .eg.go files had unnecessary imports of execgen, which is not
required in the runtime generated code, and in fact is suspicious.

Release note: None

jordanlewis commented:
> What is "IRs"?

Intermediate representation - the AST is the first representation of a program, but compilers typically transform the AST into several internal representations (like CockroachDB's Expr, opt, distsql trees) for different reasons.

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49728)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/colexec/main_test.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99HcQWCWAD0Aolcwmh:-M9AFcUV7y8M2XnuUk7c:bnjzbbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/13845e418807a3a1afe290636937f5ebb3f1e81c/pkg/sql/colexec/main_test.go#L71)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I was wondering about how one would do this at some point but ended up setting an environment variable to `false` on my machine. Turns out it was this simple, very nice change, thanks!
</blockquote></details>

:tada:
___
*[pkg/sql/colexec/execgen/inline.go, line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99LGZYEX4lYHgfvh8N:-M9A9jyt-dRxg33qMqQB:b2ytt2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline.go#L118)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I wonder why it is a block stmt, does it have to be so? Just curious. Or it doesn't matter because we're using only `body.List` below?
</blockquote></details>

It's not strictly necessary, but I find it aesthetic - it groups the related statements together. I'm okay deleting it if you want.
___
*[pkg/sql/colexec/execgen/inline.go, line 290 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99O2O-5IzpsIXvldsh:-M9AA-SE5JZ8zVzNkUeQ:b-dzd74k) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline.go#L290)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Just curious why we have `Names[0]`, can an AST node have multiple names somehow?
</blockquote></details>

`formalParam` is a `Field`, which is used in a few different places in the AST. One of the places it's used is a parameter list. In Go, you can have the syntax `func a (a, b int)`, where both `a` and `b` are ints but int is just repeated once. That is the situation in which `Names` could have more than 1 element. I'm ignoring that case for now. 

Btw, I've used https://astexplorer.net a lot to help with this stuff. It's like a Godbolt for text to ASTs.
___
*[pkg/sql/colexec/execgen/inline_test.go, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99VjaK3q9POAXYAyqr:-M9AFwZC4jKdxaKLYPlO:b5xuytt) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline_test.go#L72)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: I'd indent these `expectedRetDecls` a little bit to the right, if that's ok.
</blockquote></details>

I think they need to be just like this, because the tests respect spacing, unfortunately. I don't have it in me to fix that right now, it's so fiddly.
___
*[pkg/sql/colexec/execgen/testdata/inline, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M96KdvITnxKnqOEXInJ-r3-1:-M9AFRBC-xQyMrW4BWhO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/303646c451114b54361b8e07e43d9971e8055794/pkg/sql/colexec/execgen/testdata/inline#L1)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.
___
*[pkg/sql/colexec/execgen/testdata/inline, line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99WcFtA8L67XNSfwhB:-M9AFRpHBkk8qiyr1uHq:bl27acx) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/testdata/inline#L21)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: the indentation here seems to be done with tabs, can we adjust that? I don't have a personal opinion on tabs vs spaces, but I think in our codebase we use spaces.
</blockquote></details>

Can't adjust this unfortunately, the gofmt used by the pretty printer is hard coded to tabs. I guess I could adjust if I tried really hard but it doesn't seem worth it.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
Added another commit that gets rid of some unnecessary (and now invalid) imports of `execgen` from `colexec` generated files.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-:-M9AMGZJBuQU8wJSdxRG:b-hciv4z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rohany, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the comment about the linter.

> The .eg.go files had unnecessary imports of execgen, which is not
> required in the runtime generated code, and in fact is suspicious.

I think this became unnecessary when we recently made a change to how imports are handled. Previously, it was necessary to either surround import of `execgen` with template comments (i.e. `// {{/*` and `// */}}`) or to use it somehow (to remove unused warning).

> Reviewed 18 of 30 files at r5, 12 of 12 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-:-M9BVc7t2pCgy0PV1VnM:b-ug5mjl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

___
*[pkg/sql/colexec/select_in_tmpl.go, line 37 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M9BWDIUDVNdG5eule1G:-M9BWDIUDVNdG5eule1H:b-3lvn8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e362897a039338a2b481915b56b38d23daca57e/pkg/sql/colexec/select_in_tmpl.go#L37)):*
> ```Go
> 
> // Remove unused warnings.
> var (
> ```

super nit: could collapse it into a single line declaration.
___
*[pkg/sql/colexec/execgen/inline.go, line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99LGZYEX4lYHgfvh8N:-M9BWRQd9b_tolDj1qMZ:babjgja) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline.go#L118)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It's not strictly necessary, but I find it aesthetic - it groups the related statements together. I'm okay deleting it if you want.
</blockquote></details>

No strong preference from me.
___
*[pkg/sql/colexec/execgen/inline.go, line 290 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99O2O-5IzpsIXvldsh:-M9BWasBAy9J0rtw6LOF:bcr5s31) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline.go#L290)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`formalParam` is a `Field`, which is used in a few different places in the AST. One of the places it's used is a parameter list. In Go, you can have the syntax `func a (a, b int)`, where both `a` and `b` are ints but int is just repeated once. That is the situation in which `Names` could have more than 1 element. I'm ignoring that case for now. 

Btw, I've used https://astexplorer.net a lot to help with this stuff. It's like a Godbolt for text to ASTs.
</blockquote></details>

Got it, thanks.
___
*[pkg/sql/colexec/execgen/inline_test.go, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99VjaK3q9POAXYAyqr:-M9BWsav9Zvdde_9JsKu:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/inline_test.go#L72)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think they need to be just like this, because the tests respect spacing, unfortunately. I don't have it in me to fix that right now, it's so fiddly.
</blockquote></details>

SGTM
___
*[pkg/sql/colexec/execgen/testdata/inline, line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M99WcFtA8L67XNSfwhB:-M9BX9tCD4esGE-wKoMW:bk8t5sv) ([raw file](https://github.com/cockroachdb/cockroach/blob/257a03fd16dbe979454eb5ea0cd9b95a53afdbf0/pkg/sql/colexec/execgen/testdata/inline#L21)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Can't adjust this unfortunately, the gofmt used by the pretty printer is hard coded to tabs. I guess I could adjust if I tried really hard but it doesn't seem worth it.
</blockquote></details>

I agree, not worth it then.
___
*[pkg/testutils/lint/lint_test.go, line 1487 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-M9BXHI041KxdYUqvT3D:-M9BXHI19aBJv4WD6YgO:bxnclf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e362897a039338a2b481915b56b38d23daca57e/pkg/testutils/lint/lint_test.go#L1487)):*
> ```Go
> 			// This exception is because execgen itself uses panics during code
> 			// generation - not at execution time.
> 			":!sql/colexec/execgen/*.go",
> ```

Hm, I think we need to be more careful here because in `colexec/execgen/cmd/execgen` folder we have overloads with code snippets that will be inlined. We need to make sure that those use `colexecerror` package for panics.

Maybe something like this would work `:!sql/colexec/execgen/(^cmd)/*.go`? Alternatively, you could do `s/panic/colexecerror.InternalError/g` in the new `execgen` package and remove this exception.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
I fixed the linter (and learned something new about the syntax git uses to specify files).

jordanlewis commented:
I'll give @asubiotto a chance to take a look before merging.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> this looks great!

> Reviewed 1 of 1 files at r1, 30 of 30 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49728#-:-M9JTRbfAdQvp0tkIYxI:bdu9uj3)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @rohany, and @yuzefovich)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTRs!

bors r+
